### PR TITLE
Support for Kubernetes v1.28

### PR DESCRIPTION
**Pull request by oliver-goetz**
**How to categorize this PR?**

/area open-source
/kind enhancement

**What this PR does / why we need it**:
Support for Kubernetes v1.28.

**Which issue(s) this PR fixes**:
Part of #8291

**Special notes for your reviewer**:
- Code changes have been tested in the following scenarios:
  - Create/Delete a new (workerless) Shoot cluster with 1.28.2
  - Upgrade an existing (workerless) Shoot cluster to 1.28.2
  - Register a 1.28.2 Shoot as a Seed cluster

**Release note**:

```feature operator
Gardener can now support clusters with Kubernetes version 1.28. In order to allow creation/update of 1.28 clusters you will have to update the version of your provider extension(s) to a version that supports 1.28 as well. Please consult the respective releases and notes in the provider extension's repository.
```

```feature developer
Gardener can now support clusters with Kubernetes version 1.28. Extension developers have to prepare individual extensions as well to work with 1.28.
```

### CODE REVIEW

File: pkg/utils/version/version.go [hunk 1]
```diff
@@ -47,6 +47,10 @@ var (
 	ConstraintK8sGreaterEqual127 *semver.Constraints
 	// ConstraintK8sLess127 is a version constraint for versions < 1.27.
 	ConstraintK8sLess127 *semver.Constraints
+	// ConstraintK8sGreaterEqual128 is a version constraint for versions >= 1.28.
+	ConstraintK8sGreaterEqual128 *semver.Constraints
+	// ConstraintK8sLess128 is a version constraint for versions < 1.28.
+	ConstraintK8sLess128 *semver.Constraints
```
Comment (shafeeqes) [inline pkg/utils/version/version.go]: This is not used anywhere; do we need this? We can always add when we need it.

**Review by shafeeqes — COMMENTED** [review]
/lgtm

File: pkg/component/shootsystem/shootsystem.go [hunk 2]
```diff
@@ -305,6 +305,12 @@ func (s *shootSystem) getServiceAccountNamesToInvalidate() []string {
 			"resource-claim-controller")
 	}
 
+	if versionutils.ConstraintK8sGreaterEqual128.Check(s.values.Shoot.KubernetesVersion) {
```
Comment (ScheererJ) [inline pkg/component/shootsystem/shootsystem.go:308]: Should we not add a test for this similar to what we did in https://github.com/gardener/gardener/pull/7275/commits/ff962d055de9c84b767fa347d32f094e43f84b92?

**Review by ScheererJ — CHANGES_REQUESTED** [review]
Thanks for the change. It looks good overall.

Are the empty commits "just" markers so that the next developer working on Kubernetes v1.29 support knows what to do?

File: pkg/component/shootsystem/shootsystem.go [hunk 3]
```diff
@@ -305,6 +305,12 @@ func (s *shootSystem) getServiceAccountNamesToInvalidate() []string {
...
+		kubeControllerManagerServiceAccountNames = append(kubeControllerManagerServiceAccountNames,
+			"legacy-service-account-token-cleaner",
+			"validatingadmissionpolicy-status-controller")
```
Comment (rfranzke) [inline pkg/component/shootsystem/shootsystem.go]: Did you check whether they really appear in the `kube-system` namespace as `ServiceAccount`s?

Comment (rfranzke) [inline pkg/component/shootsystem/shootsystem.go]:
```suggestion
			"validatingadmissionpolicy-status-controller",
		)
```

Comment (oliver-goetz) [reply pkg/component/shootsystem/shootsystem.go]: Yes, when `LegacyServiceAccountTokenCleanUp` feature gate is activated for `kube-controller-manager` `legacy-service-account-token-cleaner` appears.
When `ValidatingAdmissionPolicy` feature gates is activated for `kube-controller-manager` and `admissionregistration.k8s.io/v1beta1` runtime config in `kube-apiserver` is activated `validatingadmissionpolicy-status-controller` appears.

Comment (rfranzke) [reply pkg/component/shootsystem/shootsystem.go]: Great, thank you :)

Comment (oliver-goetz) [reply pkg/component/shootsystem/shootsystem.go]: Why not `legacy-service-account-token-cleaner`?

Comment (oliver-goetz) [reply pkg/component/shootsystem/shootsystem.go]: Ah, you just want me to create a line break right? 😄

**Review by ScheererJ — COMMENTED** [review]
/lgtm

**Review by acumino — COMMENTED** [review]
Had a look together with @dimitar-kostadinov, PR looks good, we also tested the creation and deletion of the shoot with workers in the provider-local setup. 

Probably [this](https://github.com/gardener/gardener/blob/ad64150e76a4b1c7dd2e474332db73e4bc391d7b/test/framework/resources/templates/vpntunnel-copy.yaml.tpl#L31-L32) TODO can be addressed now.

From the 1.28 release notes - 
- Support for cgroup v1 is removed. (Do we need some changes for it?)
- Replaced `apiserver_storage_db_total_size_in_bytes` with `apiserver_storage_size_bytes` metric. (Can you adapt to this change?)

**Comment by oliver-goetz** [comment]
> Had a look together with @dimitar-kostadinov, PR looks good, we also tested the creation and deletion of the shoot with workers in the provider-local setup.
> 
> Probably [this](https://github.com/gardener/gardener/blob/ad64150e76a4b1c7dd2e474332db73e4bc391d7b/test/framework/resources/templates/vpntunnel-copy.yaml.tpl#L31-L32) TODO can be addressed now.

I don't know what to do here tbh 😄  
The container to which the downloaded kubectl binary is mounted [sleeps](https://github.com/gardener/gardener/blob/ad64150e76a4b1c7dd2e474332db73e4bc391d7b/test/framework/resources/templates/vpntunnel-copy.yaml.tpl#L47-L49) and `registry.k8s.io/kubectl:v1.28.2` contains the kubectl binary only.

> 
> From the 1.28 release notes -
> 
> * Support for cgroup v1 is removed. (Do we need some changes for it?)

Afaik operators have to ensure in the cloud-profiles to combine Kubernetes v1.28 with cgroupv2 host OS only.
However, when I was checking for cgroup I found `KubeletCgroupDriverFromCRI` feature gate ([ref](https://github.com/kubernetes/kubernetes/pull/118770)) which I added to the operating system config

> * Replaced `apiserver_storage_db_total_size_in_bytes` with `apiserver_storage_size_bytes` metric. (Can you adapt to this change?)

done 👍

File: pkg/component/kubeapiserver/monitoring.go [hunk 4]
```diff
@@ -26,6 +26,7 @@ import (
 
 // TODO: The etcd_object_counts metric is removed in K8s 1.25 (https://github.com/kubernetes/kubernetes/pull/110337). Remove this metric when we remove support for K8s 1.25.
 // TODO: The etcd_db_total_size_in_bytes metric is renamed to apiserver_storage_db_total_size_in_bytes in K8s 1.26 (https://github.com/kubernetes/kubernetes/pull/113310). Remove this metric when we remove support for K8s 1.26.
+// TODO: The apiserver_storage_db_total_size_in_bytes metric is replaced by apiserver_storage_size_bytes in K8s 1.28 (https://github.com/kubernetes/kubernetes/pull/118812). Remove this metric when we remove support for K8s 1.28.
```
Comment (acumino) [inline pkg/component/kubeapiserver/monitoring.go:29]:
```suggestion
// TODO: The apiserver_storage_db_total_size_in_bytes metric is replaced by apiserver_storage_size_bytes in K8s 1.28 (https://github.com/kubernetes/kubernetes/pull/118812). Remove this metric when we remove support for K8s 1.27.
```

Comment (oliver-goetz) [reply pkg/component/kubeapiserver/monitoring.go:29]: I kept this in sync with the upstream deprecation strategy on this field ([ref](https://github.com/kubernetes/kubernetes/pull/118812#discussion_r1252607835)).

File: pkg/component/shootsystem/shootsystem.go (hunk 3 above)
Comment (rfranzke) [reply pkg/component/shootsystem/shootsystem.go]: Actually yes :D

**Review by acumino — COMMENTED** [review]
/lgtm

**Pull request merged**